File: tests/Loops.cs
namespace Demo {
    public partial class ForDownToDemo {
        public static void Test() {
            for (var i = 3; i >= 1; i--) System.Console.WriteLine(i);
        }
    }
    
    public partial class ForEachExample {
        public static int Sum(int[] arr) {
            int result;
            int total;
            total = 0;
            foreach (var item in arr) total = total + item;
            result = total;
            return result;
        }
    }
    
    public partial class StepExample {
        public static int Sum() {
            int result;
            int i;
            int total;
            total = 0;
            for (i = 1; i <= 5; i += 2) total = total + i;
            result = total;
            return result;
        }
    }
    
    public partial class LoopExample {
        public static void CountThree() {
            int i;
            i = 0;
            while (true) {
                i = i + 1;
                if (i == 3) break;
            }
        }
    }
    
    public partial class RepeatExample {
        public static void LoopIt() {
            int i;
            i = 0;
            do {
                i = i + 1;
            } while (!(i > 3));
        }
    }
    
    public partial class BreakExample {
        public static void DoBreak(int[] arr) {
            int i;
            for (i = 0; i <= arr.Length - 1; i++) {
                if (arr[i] == 0) break;
                if (arr[i] < 0) continue;
            }
        }
    }
    
    public partial class WhileDemo {
        public static void Test(int[] arr) {
            int i = 0;
            while (i <= 2) i = i + arr[0];
        }
    }
    
    public partial class WhileEmptyDemo {
        public static void DoNothing() {
            int i = 0;
            while (i < 3);
        }
    }
    
    public partial class TypedForEach {
        public static int Sum(int[] arr) {
            int result;
            int total;
            total = 0;
            foreach (int item in arr) total = total + item;
            result = total;
            return result;
        }
    }
    
    public partial class ForEachIndexExample {
        public static int Sum(int[] arr) {
            int result;
            int total;
            total = 0;
            for (var idx = 0; idx < arr.Length; idx++) {
                var item = arr[idx];
                total = total + item * idx;
            }
            result = total;
            return result;
        }
    }
    
    public partial class TypedForLoop {
        public static int Sum(int[] arr) {
            int result;
            int total;
            total = 0;
            for (int i = 0; i <= length(arr) - 1; i++) total = total + arr[i];
            result = total;
            return result;
        }
    }
}

File: tests/Loops.pas
namespace Demo;

type
  ForDownToDemo = public class
  public
    class method Test();
  end;

  ForEachExample = public class
  public
    class method Sum(arr: array of Integer): Integer;
  end;

  StepExample = public class
  public
    class method Sum(): Integer;
  end;

  LoopExample = public class
  public
    class method CountThree();
  end;

  RepeatExample = public class
  public
    class method LoopIt();
  end;

  BreakExample = public class
  public
    class method DoBreak(arr: array of Integer);
  end;

  WhileDemo = public class
  public
    class method Test(arr: Array of Integer);
  end;

  WhileEmptyDemo = public class
  public
    class method DoNothing();
  end;

  WhileNoStmtDemo = public class
  public
    class method Skip();
  end;

  TypedForEach = public class
  public
    class method Sum(arr: array of Integer): Integer;
  end;

  ForEachIndexExample = public class
  public
    class method Sum(arr: array of Integer): Integer;
  end;

  TypedForLoop = public class
  public
    class method Sum(arr: array of Integer): Integer;
  end;

implementation

class method ForDownToDemo.Test();
begin
  for i := 3 downto 1 do
    System.Console.WriteLine(i);
end;

class method ForEachExample.Sum(arr: array of Integer): Integer;
var
  total: Integer;
begin
  total := 0;
  for each item in arr do
    total := total + item;
  result := total;
end;

class method StepExample.Sum(): Integer;
var
  i: Integer;
  total: Integer;
begin
  total := 0;
  for i := 1 to 5 step 2 do
    total := total + i;
  result := total;
end;

class method LoopExample.CountThree();
var
  i: Integer;
begin
  i := 0;
  loop
  begin
    i := i + 1;
    if i = 3 then break;
  end;
end;

class method RepeatExample.LoopIt();
var
  i: Integer;
begin
  i := 0;
  repeat
    i := i + 1;
  until i > 3;
end;

class method BreakExample.DoBreak(arr: array of Integer);
var
  i: Integer;
begin
  for i := 0 to arr.Length - 1 do
  begin
    if arr[i] = 0 then break;
    if arr[i] < 0 then continue;
  end;
end;

class method WhileDemo.Test(arr: Array of Integer);
var
  i: Integer := 0;
begin
  while (i <= 2) do
    i := i + arr[0];
end;

class method WhileEmptyDemo.DoNothing();
var
  i: Integer := 0;
begin
  while i < 3 do;
end;

class method WhileNoStmtDemo.Skip();
var
  i: Integer := 0;
begin
  while i < 3 do
end;

class method TypedForEach.Sum(arr: array of Integer): Integer;
var
  total: Integer;
begin
  total := 0;
  for each item: Integer in arr do
    total := total + item;
  result := total;
end;

class method ForEachIndexExample.Sum(arr: array of Integer): Integer;
var
  total: Integer;
begin
  total := 0;
  for each item in arr index idx do
    total := total + item * idx;
  result := total;
end;

class method TypedForLoop.Sum(arr: array of Integer): Integer;
var
  total: Integer;
begin
  total := 0;
  for i: Integer := 0 to length(arr) - 1 do
    total := total + arr[i];
  result := total;
end;

end.

